healthcheck/handler.go

```diff
@@ -0,0 +1,102 @@
+package HealthCheck
+
+import (
+	"encoding/json"
+	"github.com/ONSdigital/go-ns/log"
+	"net/http"
+	"time"
+)
+
+const (
+	StatusOK  = "OK"
+	StatusCritical = "CRITICAL"
+	StatusWarning  = "WARNING"
+)
+
+// Handler returns the current health check of the app
+func (hc HealthCheck) Handler(w http.ResponseWriter, req *http.Request) {
+	now := time.Now()
+	ctx := req.Context()
+
+	var checks []*Check
+	var status string
+
+	status = hc.getAppHealth()
+	for _, client := range hc.Clients {
+		if client.Check != nil{
+			checks = append(checks, client.Check)
+		}
+	}
+
+	hr := HealthResponse{
+		Status:    status,
+		Version:   hc.Version,
+		Uptime:    now.Sub(hc.StartTime),
+		StartTime: hc.StartTime,
+		Checks:    checks,
+	}
+
+	b, err := json.Marshal(hr)
+	if err != nil {
+		log.InfoCtx(ctx, "failed to marshal json", log.Data{"error": err, "HealthCheck": hr})
+		return
+	}
+
+	w.Write(b)
+
+}
+
+// isAppStartingUp returns true or false based on if each client has had results back from at least a single check
+func (hc HealthCheck) isAppStartingUp() bool{
+	for _, client := range hc.Clients {
+		if client.Check == nil {
+			return true
+		}
+	}
+	return false
+}
+
+// getAppHealth returns a status as string as to the overall current apps health based on its dependent apps health
+func (hc HealthCheck) getAppHealth() string {
+	if hc.isAppStartingUp(){
+		return StatusWarning
+	}
+	return hc.isOverallAppHealthy()
+}
+
+// isOverallAppHealthy checks every clients check for their health then produces and returns a status for this apps health
+func (hc HealthCheck)isOverallAppHealthy() string {
+	status := StatusOK
+	for _, client := range hc.Clients {
+		client.MutexCheck.Lock()
+		appHealth := hc.isCheckHealthy(client.Check)
+		client.MutexCheck.Unlock()
```

Comment:
It might be worth making this a `defer` call to ensure the lock is *always* released.

Reply:
Very good shout will add that in.